### DISCUSSION TITLE
Migration to Jolokia 2 with custom jvm-agent

### DISCUSSION THREAD

**Discussion by fbacchella** [rhuss/jolokia#630] [Q&A] [answered]
I have a projet that programmaticaly starts a JVM agent, it can be found at https://github.com/fbacchella/jrdsagent. I tried to migrate to Jolokia 2, and started a branch for that: [first jolokia 2 commit](https://github.com/fbacchella/jrdsagent/commit/6e05c9abea7a71584a0abeb3e2119841634e71e6)
There is a few differences, the LogHandler package changed, the way the server is started is slightly different.
But once it start, it fails. The following URL `http://localhost:2002/jolokia/read/jrds:type=agent/Uptime` works with Jolokia 1.7.2. But it fails with the following stack on Jolokia 2:
```
java.lang.IllegalStateException: Internal error: No request handler found for handling JmxReadRequest[attribute=Uptime, objectName = jrds:type=agent]
	at org.jolokia.server.core.backend.RequestDispatcherImpl.dispatch(RequestDispatcherImpl.java:52)
	at org.jolokia.server.core.backend.BackendManager.callRequestDispatcher(BackendManager.java:166)
	at org.jolokia.server.core.backend.BackendManager.handleRequest(BackendManager.java:93)
	at org.jolokia.server.core.http.HttpRequestHandler.executeRequest(HttpRequestHandler.java:198)
	at org.jolokia.server.core.http.HttpRequestHandler.handleGetRequest(HttpRequestHandler.java:81)
	at org.jolokia.jvmagent.handler.JolokiaHttpHandler.executeGetRequest(JolokiaHttpHandler.java:228)
	at org.jolokia.jvmagent.handler.JolokiaHttpHandler.doHandle(JolokiaHttpHandler.java:169)
	at org.jolokia.jvmagent.handler.JolokiaHttpHandler.handle(JolokiaHttpHandler.java:110)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:95)
	at jdk.httpserver/sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:82)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:98)
	at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:851)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:95)
	at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:818)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

I wonder what I missed ?

**Comment by tadayosi** [comment, marked as answer]
@fbacchella For JVM agent, use it with classifier `javaagent`. This is the standalone package with dependencies in it.
See https://repo1.maven.org/maven2/org/jolokia/jolokia-agent-jvm/2.0.0/jolokia-agent-jvm-2.0.0-javaagent.jar

**Comment by grgrzybek** [comment]
@fbacchella also, according to https://jolokia.org/migration.html#reorganized_modules you now need *three* modules at least. `jolokia-server-core` alone is not enough, you also need `jolokia-service-jmx` and `jolokia-service-serializer`.

If you use one dependency on `jolokia-agent-jvm` (without classifier) you'll get these dependencies transitively.

**Reply by fbacchella** [reply]
They are pulled:
```
$ mvn dependency:tree 
...
[INFO] +- org.jolokia:jolokia-agent-jvm:jar:2.0.0:compile
[INFO] |  +- org.jolokia:jolokia-server-core:jar:2.0.0:compile
[INFO] |  +- org.jolokia:jolokia-server-detector:jar:2.0.0:compile
[INFO] |  +- org.jolokia:jolokia-service-serializer:jar:2.0.0:compile
[INFO] |  +- org.jolokia:jolokia-service-jmx:jar:2.0.0:compile
[INFO] |  +- org.jolokia:jolokia-service-discovery:jar:2.0.0:compile
[INFO] |  +- org.jolokia:jolokia-service-history:jar:2.0.0:compile
[INFO] |  +- org.jolokia:jolokia-service-jsr160:jar:2.0.0:compile
[INFO] |  +- org.jolokia:jolokia-service-notif-pull:jar:2.0.0:compile
[INFO] |  +- org.jolokia:jolokia-service-notif-sse:jar:2.0.0:compile
[INFO] |  \- com.googlecode.json-simple:json-simple:jar:1.1.1:compile
```

They are embedded in the jar by an assembly build:
```
$ jar -tf target/jrdsagent-agentlinux.jar  | sort -uV | less
...
org/jolokia/service/jmx/
org/jolokia/service/jmx/LocalRequestHandler.class
org/jolokia/service/jmx/api/
org/jolokia/service/jmx/api/CommandHandler.class
org/jolokia/service/jmx/api/CommandHandlerManager.class
org/jolokia/service/jmx/handler/
...
org/jolokia/service/serializer/object/
org/jolokia/service/serializer/object/ArrayTypeConverter.class
org/jolokia/service/serializer/object/CompositeTypeConverter.class
org/jolokia/service/serializer/object/OpenTypeConverter.class
...
 
But when I look inside jolokia-agent-jvm-2.0.0-javaagent.jar, I see that some package were renamed:
```
org/jolokia/shaded/
org/jolokia/shaded/org/
org/jolokia/shaded/org/json/
org/jolokia/shaded/org/json/simple/
org/jolokia/shaded/org/json/simple/ItemList.class
org/jolokia/shaded/org/json/simple/JSONArray.class
org/jolokia/shaded/org/json/simple/JSONAware.class
org/jolokia/shaded/org/json/simple/JSONObject.class
org/jolokia/shaded/org/json/simple/JSONStreamAware.class
org/jolokia/shaded/org/json/simple/JSONValue.class
org/jolokia/shaded/org/json/simple/parser/
org/jolokia/shaded/org/json/simple/parser/ContainerFactory.class
org/jolokia/shaded/org/json/simple/parser/ContentHandler.class
org/jolokia/shaded/org/json/simple/parser/JSONParser.class
org/jolokia/shaded/org/json/simple/parser/ParseException.class
org/jolokia/shaded/org/json/simple/parser/Yylex.class
org/jolokia/shaded/org/json/simple/parser/Yytoken.class
```

I will try to do that in my own assembly to see if it helps.

**Comment by grgrzybek** [comment]
`org.jolokia:jolokia-agent-jvm` is fine - it has transitive dependencies on several other Jolokia libraries and json-simple.

However if you use **2nd** variant of `org.jolokia:jolokia-agent-jvm` (the one with `javaagent` classifier), you'll both get the transitive dependencies and also everything inside (which is simply everything + json-simple with shaded package)

so either you use:
```xml
<dependency>
  <groupId>org.jolokia</groupId>
  <artifactId>jolokia-agent-jvm</artifactId>
</dependency>
```

or:
```xml
<dependency>
  <groupId>org.jolokia</groupId>
  <artifactId>jolokia-agent-jvm</artifactId>
  <classifier>javaagent</classifier>
  <exclusions>
    <exclusion>
      <groupId>*</groupId>
      <artifactId>*</artifactId>
    <exclusion>
  </exclusions>
</dependency>
```

**Reply by fbacchella** [reply]
I don’t think the first solution works. See https://github.com/jolokia/jolokia/blob/56796330e4ad084e8a14b0d7523f18ebbd79b0a4/agent/jvm/pom.xml#L156. Classes are renamed. So the second solution is the best, I tested it, it works.

**Reply by grgrzybek** [reply]
See that `<shadedClassifierName>javaagent</shadedClassifierName>` is configured, so you have two artifacts:
* https://repo1.maven.org/maven2/org/jolokia/jolokia-agent-jvm/2.0.0/jolokia-agent-jvm-2.0.0.jar
* https://repo1.maven.org/maven2/org/jolokia/jolokia-agent-jvm/2.0.0/jolokia-agent-jvm-2.0.0-javaagent.jar
